portal/src/main/webapp/js/api/cbioportal-datamanager.js

```diff
@@ -1437,6 +1455,49 @@ window.initDatamanager = function (genetic_profile_ids, oql_query, cancer_study_
 	'getPatientHeatmapData': function(genetic_profile_id, genes) {
 	    return getHeatmapDataCached(this, genetic_profile_id, genes, 'patient');
 	},
+	'getClusteringOrder': function (case_ui_map, track_uid_map, heatmapData, case_ids) {
+			var study_id = QuerySession.getCancerStudyIds()[0];
+			var def = new $.Deferred();
+			//prepare input:
+			var cluster_input = {};
+			for (var i = 0; i < case_ids.length; i++) {
+				//case ids as key:
+				cluster_input[case_ids[i]] = {};
+				//iterate over genetic entities and get the sample data (heatmap data has genetic entityId as key):
+				for (var j = 0; j < heatmapData.length; j++) {
+					var entityId = heatmapData[j].gene;
+					//small validation/defensive programming:
+					if (!entityId) {
+						throw new Error("Unexpected error during getClusteringOrder: attribute 'gene' not found");
+					}
+					var value = heatmapData[j].oncoprint_data[i].profile_data;
+					cluster_input[case_ids[i]][entityId] = value;
+				}					
+			}
+			//do hierarchical clustering in background:
+			$.when(cbio.stat.hclusterCases(cluster_input), cbio.stat.hclusterGeneticEntities(cluster_input)).then(
+					function (sampleClusterOrder, entityClusterOrder) {
```

Comment:
long story. But here it makes sense as we will have heatmaps for other non-gene entities (like proteins or genesets). To keep this code generic, I chose to use a generic naming here.